### DISCUSSION TITLE
Why processGenericAccessTokenResponse never validate the id_token signature ?

### DISCUSSION THREAD

**Discussion by kant-un** [panva/oauth4webapi#60] [Q&A] [answered]
Hello,
could you tell me if it's intended that the the _processGenericAccessTokenResponse_ never validates the id_token **signature** ?
_validateJwt_ is indeed always called with noSignatureCheck in this function.

I am using _processAuthorizationCodeOpenIDResponse_ (that calls _processGenericAccessTokenResponse_) and i was hoping that it would also validate the signature but it seems that it's not possible.

The other way to validate the id_token would be by using _validateJwtAuthResponse_ but it looks like it's only for JARM flow for which a 'response' parameter is needed. The problem is that i don't have a 'response' param in my flow (PKCE with only code and state in the callback) so the function throws an error. 

I saw that someone mentioned another package 'jose' that have the JWKS validation. Should I use this instead or is it planned to have a way to validate an id_token in _processAuthorizationCodeOpenIDResponse_ ?

Thanks for your time

**Comment by panva** [comment, marked as answer]
> could you tell me if it's intended that the ... never validates the id_token signature?

It is intended. It is possible to do so because the only flows that are supported by this module are ones where an ID Token is retrieved from an https (TLS protected) endpoint and the specification does not require the client to check the signature of the ID Token in such cases.

This behaviour is also certified by the OpenID Connect Certification Program.

> If the ID Token is received via direct communication between the Client and the Token Endpoint (which it is in this flow), the TLS server validation MAY be used to validate the issuer in place of checking the token signature.

[source](https://openid.net/specs/openid-connect-core-1_0.html#IDTokenValidation:~:text=If%20the%20ID%20Token%20is%20received%20via%20direct%20communication%20between%20the%20Client%20and%20the%20Token%20Endpoint%20(which%20it%20is%20in%20this%20flow)%2C%20the%20TLS%20server%20validation%20MAY%20be%20used%20to%20validate%20the%20issuer%20in%20place%20of%20checking%20the%20token%20signature.)

For all intents and purposes, when you retrieve the ID Token from this module you don't need to validate it futher, you can immediately use [getValidatedIdTokenClaims](https://github.com/panva/oauth4webapi/blob/v2.2.1/docs/functions/getValidatedIdTokenClaims.md) on the response and get the claims that were already in it.

If you want to be checking the ID Token on any subsequent use, for instance followup requests after it has maybe been exposed to users, by all means, use a library meant for validating JWTs to ensure it was not tampered with.

**Reply by kant-un** [reply]
Thanks a lot for your response.
You are correct, I don't have to validate the token in this flow according to the source you provided.
I am receiving the id_token in a Single Page Application. I though I'd have to validate it but it is indeed protected by the TLS connection to the server.
I only need to validate the id_token in my backend/apis server as it might be tempered with as you said.
Thanks again !
